pil-analyzer/src/evaluator.rs

```diff
@@ -505,6 +505,14 @@ impl<'a, T: FieldElement> SymbolLookup<'a, T> for Definitions<'a> {
     fn lookup_public_reference(&self, name: &str) -> Result<Arc<Value<'a, T>>, EvalError> {
         Ok(Value::from(AlgebraicExpression::PublicReference(name.to_string())).into())
     }
+
+    fn eval_expr(&self, expr: &AlgebraicExpression<T>) -> Result<Arc<Value<'a, T>>, EvalError> {
```

Comment:
Seems generic enough to go into the trait itself. But still, I'm wondering if this is conceptually correct to call `eval` outside of `constr` (i.e. condenser) context, even on numbers.

Reply:
Yeah this is still a hack, I think complex expressions like `3 + 4` would still fail. I wonder if I'd still have that problem following your approach in #1299, because maybe then a different `SymbolLookup` would be used? Will look into that when polishing the PR.

Reply:
It seems like there should a way to cast from `expr` to `fe` (or `int`) which fails unless the expression contains anyhing but a number, binary operation, or unary operation.

Reply:
Fixed by rebasing onto #1346